Front-end-landing-page

ijromerob opened the pull request:
## Why was this PR opened?

## Brief Description
I created the CoursesContext, the Routes, and the LandingPage.
The landing page has a search functionality to show the card that is needed.

## Unrelated Changes

## UI Screenshots

## Link to Referenced Trello Cards

## Mentions

ijromerob committed:
created courses context and added it in main. Created landing page

ijromerob committed:
created app routes, updated the way to get to login page

ijromerob committed:
added search functionality and reset to courses

Hoorasadat commented on src/App.jsx:
```diff
@@ -2,28 +2,23 @@
+    <>
+      <NavBar />
+      <AppRoutes />
+      {/* {userInfo == null ? <LoginPage /> : <HomePage />} */}
```
Why you have commented this line out?

ijromerob replied:
This was part of the old process before the AppRoutes was created. I will delete it. Thanks for mentioning it

Hoorasadat commented on src/AppRoutes.jsx:
```diff
@@ -0,0 +1,28 @@
+import LandingPage from "./pages/LandingPage";
+import HomePage from "./pages/HomePage";
+import LoginPage from "./pages/LoginPage";
+import { useContext } from "react";
```
Since useContext is imported from react, you can combine it with the React import.

ijromerob replied:
Thank you for pointing this out! That was an excelent observation

Hoorasadat commented on src/AppRoutes.jsx:
```diff
@@ -0,0 +1,28 @@
+      <Routes>
+        <Route
+          path="/"
+          element={userInfo == null ? <LandingPage /> : <Navigate to="/homepage" />}
```
You're using userInfo == null in some places and userInfo in others. It’s better to use a consistent approach throughout, either checking explicitly for null or using a truthy/falsy check.

ijromerob replied:
Thanks for the suggestions. I have made the change

Hoorasadat commented on src/AppRoutes.jsx:
```diff
@@ -0,0 +1,28 @@
+  return (
+    <BrowserRouter>
+      <Routes>
+        <Route
```
You might want to add a * route for handling undefined paths, possibly redirecting to a "Not Found" page or the homepage.

ijromerob replied:
As suggested, I added the *route

Hoorasadat commented on src/AppRoutes.jsx:
```diff
@@ -0,0 +1,28 @@
+  );
+}
+
+export default AppRoutes;
```
It’s usually a good practice to wrap the entire app with BrowserRouter in a higher-level component, like index.js or App.js, to avoid re-rendering the router when switching routes. This also allows easier integration with other components like context providers.

ijromerob replied:
Thanks for the comment, I will move it to main.jsx

Hoorasadat commented on src/AppRoutes.jsx:
```diff
@@ -0,0 +1,28 @@
+      <Routes>
+        <Route
+          path="/"
+          element={userInfo == null ? <LandingPage /> : <Navigate to="/homepage" />}
```
If userInfo is not null, it might still be beneficial to check if it’s an empty object or another falsy value, depending on how userInfo is structured.

Hoorasadat commented on src/contexts/CoursesContext.jsx:
```diff
@@ -0,0 +1,25 @@
+UserContextProvider.propTypes = {
+  children: PropTypes.object
+};
+
```
You are assigning propTypes to UserContextProvider instead of CoursesContextProvider. This seems to be a typo.

ijromerob replied:
You are right!

Hoorasadat commented on src/contexts/CoursesContext.jsx:
```diff
@@ -0,0 +1,25 @@
+}
+
+UserContextProvider.propTypes = {
+  children: PropTypes.object
```
Instead of using PropTypes.object, it’s better to use PropTypes.node, which is more accurate for React children since it can include everything that can be rendered, not just objects.

ijromerob replied:
I agree it can add more flexibility

Hoorasadat commented on src/pages/LandingPage.jsx:
```diff
@@ -0,0 +1,84 @@
+  const [searchText, setSearchText] = useState("");
+  // state for the courses that will alter the courses displayed
+  const courses = coursesInfo?.Courses_data ? coursesInfo.Courses_data : [];
+  const [coursesDisplay, setCoursesDisplay] = useState(courses);
```
coursesDisplay should be initialized based on courses directly. You might want to use useEffect to update coursesDisplay whenever coursesInfo changes.

ijromerob replied:
that's a great suggestion I will add the useEffect with coursesInfo as a dependency

Hoorasadat commented on src/pages/LandingPage.jsx:
```diff
@@ -0,0 +1,84 @@
+
+  // FUNCTIONS
+  function searchInCourses(searchString) {
+    const filteredCourses = coursesDisplay.filter((c) => {
```
Consider trimming and lowercasing the search string and course fields for a case-insensitive search.
You are currently filtering coursesDisplay, which is already filtered. Instead, filter courses directly to always base it on the complete list of courses.

ijromerob replied:
Great observation, I have added the trimming and lowercase. Also the for each title and description was made this adjustment

Hoorasadat commented on src/pages/LandingPage.jsx:
```diff
@@ -0,0 +1,84 @@
+  // FUNCTIONS
+  function searchInCourses(searchString) {
+    const filteredCourses = coursesDisplay.filter((c) => {
+      return c.title.includes(searchString) || c.description.includes(searchString);
```
Consider handling the case when there are no courses to display or when the search returns no results.

ijromerob replied:
as per this case, the lenght of the array would be 0. From now on the lenght will be checked to display a message when this criteria is met

Hoorasadat commented on src/pages/LandingPage.jsx:
```diff
@@ -0,0 +1,84 @@
+  // Course has the information for the course that would like to be displayed
+  const [course, setCourse] = useState(null);
+  // controlled input state that will help to filter the courses that are displayed
+  const [searchText, setSearchText] = useState("");
```
Ensure that searchText is reset or cleared when appropriate, such as after performing a search or resetting the courses.

ijromerob replied:
after performing a search the text is left there as an indication to the user of what are they looking at.
I made some changes to ensure that after reseting the search, the input is cleared.

Hoorasadat reviewed (commented):
Thanks for the code. I have added some comments. Hope they are useful!

ijromerob committed:
Standardized imports according to best practices.

Change the AppRoutes to include  * cases

Changed the propTypes in context with more flexibility making sure it accepts .node

Improved search feature to enhance discoverability
